extension/runtime/src/main/java/io/quarkiverse/infinispan/embedded/runtime/cache/InfinispanCacheImpl.java

```diff
@@ -0,0 +1,220 @@
+                                        }
+                                        computationResults.remove(key);
+                                    });
+                        }
+                    });
+            return resultAsync;
+        })).emitOn(new Executor() {
+            // We need make sure we go back to the original context when the cache value is computed.
+            // Otherwise, we would always emit on the context having computed the value, which could
+            // break the duplicated context isolation.
+            @Override
+            public void execute(Runnable command) {
+                Context ctx = Vertx.currentContext();
+                if (context == null) {
+                    // We didn't capture a context
+                    if (ctx == null) {
+                        // We are not on a context => we can execute immediately.
+                        command.run();
+                    } else {
+                        // We are on a context.
+                        // We cannot continue on the current context as we may share a duplicated context.
+                        // We need a new one. Note that duplicate() does not duplicate the duplicated context,
+                        // but the root context.
+                        ((ContextInternal) ctx).duplicate()
+                                .runOnContext(new Handler<Void>() {
+                                    @Override
+                                    public void handle(Void ignored) {
+                                        command.run();
+                                    }
+                                });
+                    }
+                } else {
+                    // We captured a context.
+                    if (ctx == context) {
+                        // We are on the same context => we can execute immediately
+                        command.run();
+                    } else {
+                        // 1) We are not on a context (ctx == null) => we need to switch to the captured context.
+                        // 2) We are on a different context (ctx != null) => we need to switch to the captured context.
+                        context.runOnContext(new Handler<Void>() {
+                            @Override
+                            public void handle(Void ignored) {
+                                command.run();
+                            }
+                        });
+                    }
+                }
+            }
+        });
+    }
+
+    @Override
+    public Uni<Void> invalidate(Object key) {
+        return Uni.createFrom().completionStage(() -> cache.removeAsync(key));
+    }
+
+    @Override
+    public Uni<Void> invalidateAll() {
+        return Uni.createFrom().completionStage(() -> cache.clearAsync());
+    }
+
+    @Override
+    public Uni<Void> invalidateIf(Predicate<Object> predicate) {
+        //TODO: ASK will best strategy
```

Comment:
Maybe change this to mention that this can be implemented in a nicer fashion when on ISPN 15.2+